### ISSUE TITLE
support a DHT-only mode

### ISSUE THREAD

**Issue by willscott** [ipfs-shipyard/someguy#13]
In thinking about what a joint deployment of indexer & dht might look like, we likely want to run instances of accelerated DHT clients that can fill in DHT records in delegated routing responses.

We probably want those responses to take advantage of the same caching as current indexing responses.

The potential difference in latency means that for uncached responses we'll likely want to make sure we split the response into multiple response packets as results return, and parse incrementally on the client side.

**Comment by lidel** [comment]
If we want DHT-only proxy mode that allows us shutting down DHT client on gateways, and delegating  it to `someguy` instance deployed in or next to cid.contact, we need:

1. https://github.com/ipfs-shipyard/someguy/issues/12
2. https://github.com/ipfs-shipyard/someguy/issues/14
3. https://github.com/ipfs-shipyard/someguy/issues/15

**Comment by willscott** [comment]
We're part way through the cascade of the direct cascade for `find-providers` via https://github.com/ipni/caskadht

The IPNS part would be, i think, on the top level ipfs.tech gateways, rather than something needed in the L1s in my current understanding of architecture. does that match with your understanding?

**Comment by masih** [comment]
@lidel would it make sense to move the three issues above to `caskadht`? That's the service we plan to run on cid.contact.

**Comment by lidel** [comment]
@willscott depends how Block/CAR requests are handled.
* If CAR request for `/ipns/{id}` is sent directly to L1, then L1 needs to be able to resolve it to a CID.
* if there is no HTTP API for IPNS delegation, then `ipfs.io` gateway  still needs to run DHT client to resolve IPNS (which defeats the plan to remove DHT cost and move it to saturn/ipni)

@masih will `caskadht` be exposing streaming API compatible with [IPIP-337](https://github.com/ipfs/specs/pull/337) (`GET /routing/v1/providers/{CID}`)?  If `caskadht` can do that, then no longer need `someguy` :+1:

**Comment by masih** [comment]
> will caskadht be exposing streaming API compatible with IPIP-337

Yes. That code already implements ndjosn response. I'll add HTTP delegated routing APIs next.

**Comment by lidel** [comment]
Hm.. we need a binary that has only one API, the IPIP-337 one. 
Otherwise, we will be leaking internal proprietary APIs to IPFS ecosystem. We have too many APIs doing the same thing, it kills UX/DX.

**Comment by willscott** [comment]
> Hm.. we need a binary that has only one API, the IPIP-337 one.

Do you mean as what we run on cid.contact? I'm confused about this requirement - indexers have compatibility back to reframe and previous delegated routing APIs, and will retire those as we see usage drop off.
Making a separate container here adds operational complexity and doesn't seem like it changes that.

**Comment by lidel** [comment]
> my sense is that there are relatively few IPNS lookups, so that even with IPNS resolution remaining gateways, we would still mostly remove gateway costs.

Yes, just flagging that adding DHT server to new ipfs.io adds to the scope on our end, 
and still does not remove IPNS resolution requirement from L1s.

> Do you mean as what we run on cid.contact?

No, this is for IPFS Ecosystem.  What people can run themselves.

Whatever IPFS Stewards build, users are able to self-host, and remove their dependency on infrastructure run by a third-party such as Protocol Labs.   In this case, DHT proxy compatible with IPIP-337 is a very useful component which we want to provide for download at https://dist.ipfs.tech   (just like we do with libp2p-relay-daemon).

**Comment by willscott** [comment]
> Reusable DHT proxy that speaks IPIP-337 + ndjson.

this is probably best served by something like someguy rather than caskadht.

I'm happy to defer to @masih  - but if the expectation of an ecosystem delegation is that it can't include non-ipip behavior, that will likely create tensions with service deployment goals.

**Comment by masih** [comment]
> Reusable DHT proxy that speaks IPIP-337 + ndjson.

I am easy. I am happy to make `caskadht` be that, and or move the repo wherever you see fit. Also happy for `someguy` to be it. Considering `caskadht` implements `ndjson`, probably using that is the path of less resistance.

**Comment by lidel** [comment]
Ack, I think focusing on `caskadht` is fine for the immediate task at hand (especially if it is being hidden behind cid.contact, and that is the only IPNI in Kubo).

Long term, for `delegate.routing.ipfs.tech` (https://github.com/protocol/bifrost-infra/issues/2142)  we will need productized `someguy`, something that can be put in front of multiple  IPNIs like cid.contact, and also run the own DHT, so the lookups in self-hosted setups and our browser nodes (helia, js-ipfs) are not tied to well-being of a specific IPNI.

**Comment by willscott** [comment]
I would hope long term we can use ambient discover as the way to not depend on well-being of a specific IPNI. `delegate.routing.ipfs.tech` will have similar single-point-of-failure characteristics as a single IPNI.
